src/runCompletion.ts

```diff
@@ -51,6 +76,16 @@ export default async function runCompletion({
     indentation_size: getTabSize(),
     sdk_path: getSDKPath(document.languageId),
   };
+  const notebookEditor = vscode.window.activeNotebookEditor;
+
+  if (notebookEditor) {
+    const { before, after } = calculateContextForJupyterNotebook(
+      notebookEditor,
+      document.uri
+    );
+    requestData.before = [before, requestData.before].join("\n");
```

Comment:
i'd prefer an approach that's more "immutable". how about introducing a `getContextParts` function that returns something like `{before: string, after: string}`? this function could calculate the 'before' and 'after' based on whether it's a notebook or a regular document, instead of directly mutating the requestData. WDYT?